https://issues.redhat.com/browse/ACM-15508 - provide more info about extraLabels for SiteConfig (#7243)
open-cluster-management/rhacm-docs: mce_acm_integration/siteconfig/siteconfig_installation_templates.adoc

The {sco} applies your custom labels to the manifests that you specify in the `ClusterInstance` custom resource.

When creating the labels, you must specify a manifest type to allow the {sco} to apply your custom labels to the appropriate manifest.
However, the labels are arbitrary and you can set any key and value pairs.

.Example application of extraLabels
[source,yaml]
----
apiVersion: siteconfig.open-cluster-management.io/v1alpha1
kind: ClusterInstance
metadata:
  name: "example-sno"
  namespace: "example-sno"
spec:
  [...]
  clusterName: "example-sno"
  extraLabels: <1>
    ManagedCluster:
      common: "true"
      group-du: ""
      sites : "example-sno"
      vendor: "auto-detect"
  nodes:
    - hostName: "example-sno.example.redhat.com"
      role: "master"
      extraLabels: <2>
        BareMetalHost:
          "testExtraLabel": "success"
----
<1> These are cluster-level labels that the {sco} applies to the the `ManagedCluster` manifest.
<2> This is a node-level label that the {sco} applies to the `BareMetalHost` manifest.

.Example applied labels

> [source,json]

Reviewer: Perhaps provide the `oc` command used to obtain the `ManagedCluster` label output?

----
{
  "cluster.open-cluster-management.io/clusterset": "default",
  "common": "true",
  "group-du": "",
  "name": "example-sno",
  "siteconfig.open-cluster-management.io/owned-by": "example-sno_example-sno",
  "sites": "example-sno",

>   "vendor": "auto-detect"

Reviewer: The following lines can be omitted from the sample output:
```
"cluster.open-cluster-management.io/clusterset": "default",
"siteconfig.open-cluster-management.io/owned-by": "example-sno_example-sno",
```
These can be replaced with 
```
 "common": "true",
  "group-du": "",
  etc
   ...
```

}
----

> [source,json]

Reviewer: Again, let's add the `oc` command used to generate the output below.

----
{
  "infraenvs.agent-install.openshift.io": "example-sno",

>   "siteconfig.open-cluster-management.io/owned-by": "example-sno_example-sno",

Reviewer: These two lines can be omitted and replaced with ellipses.

  namespace: example-namespace
----

If the {sco} deletes the resources, `KlusterletAddonConfig` and `ManagedCluster` custom resources are the first to be deleted, while the `AgentClusterInstall` and `ClusterDeployment` custom resources are the last.

[#extra-labels]

> == Configuration of custom labels and annotations

Reviewer: ```suggestion
== Configuration of additional annotations and labels
```
I think we can sort the elements alphabetically, i.e. annotations and then labels.

  namespace: example-namespace
----

If the {sco} deletes the resources, `KlusterletAddonConfig` and `ManagedCluster` custom resources are the first to be deleted, while the `AgentClusterInstall` and `ClusterDeployment` custom resources are the last.

[#extra-annotations-labels]
== Configuration of additional annotations and labels

> You can configure additional additional annotations and labels to both cluster-level and node-level installation manifests by using the `extraLabels` and `extraAnnotations` fields in the `ClusterInstance` API.

Reviewer: ```suggestion
You can configure additional annotations and labels to both cluster-level and node-level installation manifests by using the `extraAnnotations` and `extraLabels` fields in the `ClusterInstance` API.
```

The {sco} applies your additional annotations and labels to the manifests that you specify in the `ClusterInstance` additional resource.

> When creating your additional annotations and labels, you must specify a manifest type to allow the {sco} to apply your them to the appropriate manifest.

Reviewer: Can we fix this sentence: `... to apply your them to the appropriate manifest.`

> However, the annotations and labels are arbitrary and you can set any key and value pairs.

Reviewer: ```suggestion
However, the annotations and labels are arbitrary and you can set any key and value pairs that are meaningful to your application.
```

> .Example application of extraLabels

Reviewer: ```suggestion
.Example application of extraAnnotations and extraLabels
```

  clusterName: "example-sno"
  extraAnnotations: <1>
    ClusterDeployment:
      myClusterAnnotation: success
  extraLabels: <1>
    ManagedCluster:
      common: "true"
      group-du: ""
  nodes:
    - hostName: "example-sno.example.redhat.com"
      role: "master"
      extraAnnotations: <2>
        BareMetalHost:
          myNodeAnnotation: success
      extraLabels: <2>
        BareMetalHost:
          "testExtraLabel": "success"
----
<1> These are cluster-level annotations and labels that the {sco} applies to the the `ManagedCluster` and `ClusterDeployment` manifests.
<2> These are node-level annotations and labels that the {sco} applies to the `BareMetalHost` manifest.

. You can verify that your additional labels are applied by running the following command:
+
--
[source,terminal]
----
$ oc get managedclusters example-sno -ojsonpath='{.metadata.labels}' | jq
----

[source,json]
----
{
  "common": "true",
  "group-du": "",
  ...
}
----
--

. You can verify that your additional annotations are applied by running the following command:
+
--
[source,terminal]
----
$ oc get bmh example-sno.example.redhat.com -n example-sno -ojsonpath='{.metadata.annotations}' | jq
----

> .Example applied labels

Reviewer: Does this line `.Example applied labels` belong somewhere else perhaps above the line?
```
. You can verify that your additional labels are applied by running the following command:
```

  namespace: example-namespace
----

If the {sco} deletes the resources, `KlusterletAddonConfig` and `ManagedCluster` custom resources are the first to be deleted, while the `AgentClusterInstall` and `ClusterDeployment` custom resources are the last.

[#extra-annotations-labels]
== Configuration of additional annotations and labels

> You can configure additional additional annotations and labels to both cluster-level and node-level installation manifests by using the `extraAnnotations` and `extraLabels` fields in the `ClusterInstance` API.

Reviewer 2: 'additional' is duplicated here

Author: Thanks, updated!

Reviewer: Looks like it was not addressed here https://github.com/stolostron/rhacm-docs/pull/7243#discussion_r1866208457.

The {sco} applies your additional annotations and labels to the manifests that you specify in the `ClusterInstance` additional resource.

When creating your additional annotations and labels, you must specify a manifest type to allow the {sco} to apply them to the appropriate manifest.

> However, the annotations and labels are arbitrary and you can set any key and value pairs that are meaningful to your applications.

Reviewer 2: I like giving the first order label a name of 'manifest type'.

I don't think we need to point out that the annotations and labels themselves are arbitrary, but it is useful to point out that the 'manifest type' itself is an arbitrary label. Especially since all the examples use the `Kind` of the targeted manifest as the 'manifest type' label, but this is only a convention.

Reviewer: Note that the manifest type is not an arbitrary label in this context. The manifest type specifies the resource `kind` for which annotations / labels must be applied to. However, the manifest type in the installation templates can be an arbitrary label - this though is unrelated.

Author: My understanding is if the first order label doesn't match the target manifest type, the SiteConfig operator won't apply the annotation/label to the intended manifest(s)

```  
extraAnnotations: 
  ClusterDeployment: #this must match the target manifest type, this is the resource that the SiteConfig operator applies the annotation to
    myClusterAnnotation: success #this is the annotation that'll be applied, it can be whatever you'd like
```

Same for labels.

Reviewer 2: That is important to clear up, thanks for the explanation, @sakhoury and @amolnar-rh . In this case, the documentation as provided is good as it appears in this PR.

> However, the annotations and labels are arbitrary and you can set any key and value pairs that are meaningful to your applications.
> 
> .Example application of extraAnnotations and extraLabels
> [source,yaml]

Reviewer 3: I like how you created a title that explains what the sample is. However, we usually prompt the user for the example by saying the following statement: 

View the following example application of `extraAnnotations` and `extraLabels`:
```suggestion
However, the annotations and labels are arbitrary and you can set any key and value pairs that are meaningful to your applications.

.Example application of extraAnnotations and extraLabels
[source,yaml]
```

----
apiVersion: siteconfig.open-cluster-management.io/v1alpha1
kind: ClusterInstance
metadata:
  name: "example-sno"
  namespace: "example-sno"
spec:
  [...]
  clusterName: "example-sno"
  extraAnnotations: <1>
    ClusterDeployment:
      myClusterAnnotation: success
  extraLabels: <1>
    ManagedCluster:
      common: "true"
      group-du: ""
  nodes:
    - hostName: "example-sno.example.redhat.com"
      role: "master"
      extraAnnotations: <2>
        BareMetalHost:
          myNodeAnnotation: success
      extraLabels: <2>
        BareMetalHost:
          "testExtraLabel": "success"
----

> <1> These are cluster-level annotations and labels that the {sco} applies to the the `ManagedCluster` and `ClusterDeployment` manifests.
> <2> These are node-level annotations and labels that the {sco} applies to the `BareMetalHost` manifest.

Reviewer 3: ```suggestion
<1> This field supports cluster-level annotations and labels that the {sco} applies to the the `ManagedCluster` and `ClusterDeployment` manifests.
<2> This field supports node-level annotations and labels that the {sco} applies to the `BareMetalHost` manifest.
```

> . You can verify that your additional labels are applied by running the following command:
> +
> --
> [source,terminal]
> ----
> $ oc get managedclusters example-sno -ojsonpath='{.metadata.labels}' | jq
> ----

Reviewer 3: Is it your intention to make this a step? From my perspective this seems like a concept section and not a task. What do the two dashes do ? We also typically remove the dollar sign in the commands. 

```suggestion
. You can verify that your additional labels are applied by running the following command:

+
[source,terminal]
----
oc get managedclusters example-sno -ojsonpath='{.metadata.labels}' | jq
----
```

Author: It is a concept section but we need to include how to verify that the labels/annotations are applied. Applying extraAnnotations/extraLabels is not part of any of the procedures so not sure how to better structure it :/

> .Example applied labels

Reviewer 3: Similar comment, I like how you use the alt text to describe the YAML, but we typically introduce the example to the user. 

"View the following example of applied labels:"

[source,json]
----
{
  "common": "true",
  "group-du": "",
  ...
}
----

> --

Reviewer 3: Not sure what the dashes are for 
```suggestion
```

Author: Sorry, I was jumping between ACM and OCP PRs, I'm not even sure this works in the ACM build.

Btw, it groups together paragraphs/code snippets that would have different level of indentation without it, which would make it look like one is in the next step. I often have to use the dashes instead of the `+` signs when they just stop working. I find it's much more reliable than playing around with the `+`. I attached two pictures from my local preview, hopefully that demonstrates it better :)

- Without dashes:
![Screenshot From 2024-12-05 14-30-07](https://github.com/user-attachments/assets/84bc8c7e-f2ff-4311-a174-e761fed16ae3)

- With dashes:
![Screenshot From 2024-12-05 14-30-35](https://github.com/user-attachments/assets/ddb9fc8f-8bc7-436a-a4ea-0ba086b043ca)

> . You can verify that your additional annotations are applied by running the following command:
> +
> --
> [source,terminal]
> ----
> $ oc get bmh example-sno.example.redhat.com -n example-sno -ojsonpath='{.metadata.annotations}' | jq
> ----
> 
> .Example applied annotation
> [source,json]
> ----
> {
>   "myNodeAnnotation": "success",
>   ...
> }
> ----
> --

Reviewer 3: ```suggestion
. You can verify that your additional annotations are applied by running the following command:

+
[source,terminal]
----
oc get bmh example-sno.example.redhat.com -n example-sno -ojsonpath='{.metadata.annotations}' | jq
----

.Example applied annotation
[source,json]
----
{
  "myNodeAnnotation": "success",
  ...
}
----
```

  namespace: example-namespace
----

If the {sco} deletes the resources, `KlusterletAddonConfig` and `ManagedCluster` custom resources are the first to be deleted, while the `AgentClusterInstall` and `ClusterDeployment` custom resources are the last.

[#extra-annotations-labels]
== Configuration of additional annotations and labels

You can configure additional annotations and labels to both cluster-level and node-level installation manifests by using the `extraAnnotations` and `extraLabels` fields in the `ClusterInstance` API.

> The {sco} applies your additional annotations and labels to the manifests that you specify in the `ClusterInstance` additional resource.

Reviewer: Can this be rephrased?
> The {sco} applies your `additional` ... in the ClusterInstance `additional` resource.

Author: Updated!

  namespace: example-namespace
----

If the {sco} deletes the resources, `KlusterletAddonConfig` and `ManagedCluster` custom resources are the first to be deleted, while the `AgentClusterInstall` and `ClusterDeployment` custom resources are the last.

[#extra-annotations-labels]
== Configuration of additional annotations and labels

You can configure additional annotations and labels to both cluster-level and node-level installation manifests by using the `extraAnnotations` and `extraLabels` fields in the `ClusterInstance` API.
The {sco} applies your additional annotations and labels to the manifests that you specify in the `ClusterInstance` resource.

When creating your additional annotations and labels, you must specify a manifest type to allow the {sco} to apply them to all the matching manifests.
However, the annotations and labels are arbitrary and you can set any key and value pairs that are meaningful to your applications.

*Note:* The additional annotations and labels are only applied to the resources that were rendered through the referenced templates.

> View the following example application of extraAnnotations and extraLabels:

Reviewer 3: ```suggestion
View the following example application of `extraAnnotations` and `extraLabels`:
```

.Example application of extraAnnotations and extraLabels
[source,yaml]
----
apiVersion: siteconfig.open-cluster-management.io/v1alpha1
kind: ClusterInstance
metadata:
  name: "example-sno"
  namespace: "example-sno"
spec:
  [...]
  clusterName: "example-sno"
  extraAnnotations: <1>
    ClusterDeployment:
      myClusterAnnotation: success
  extraLabels: <1>
    ManagedCluster:
      common: "true"
      group-du: ""
  nodes:
    - hostName: "example-sno.example.redhat.com"
      role: "master"
      extraAnnotations: <2>
        BareMetalHost:
          myNodeAnnotation: success
      extraLabels: <2>
        BareMetalHost:
          "testExtraLabel": "success"
----
<1> This field supports cluster-level annotations and labels that the {sco} applies to the the `ManagedCluster` and `ClusterDeployment` manifests.
<2> This field supports node-level annotations and labels that the {sco} applies to the `BareMetalHost` manifest.

. You can verify that your additional labels are applied by running the following command:
+
[source,terminal]
----
oc get managedclusters example-sno -ojsonpath='{.metadata.labels}' | jq

> ----

Reviewer 3: Thanks for explaining why you use the two dashes. It's best to skip a line for the command to render correctly. As it relates to the step, maybe we can use a bullet point instead of an ordered step? What are your thoughts?
```suggestion
. You can verify that your additional labels are applied by running the following command:

+
[source,terminal]
----
oc get managedclusters example-sno -ojsonpath='{.metadata.labels}' | jq
----
```

Author: I like it, updated it with bullet points!

----

View the following example of applied labels:

.Example applied labels
[source,json]
----
{
  "common": "true",
  "group-du": "",
  ...
}
----

* You can verify that your additional annotations are applied by running the following command:

+
[source,terminal]
----
oc get bmh example-sno.example.redhat.com -n example-sno -ojsonpath='{.metadata.annotations}' | jq
----

View the following example of applied annotations:

.Example applied annotation
[source,json]
----
{
  "myNodeAnnotation": "success",
  ...
}

> ----

Reviewer 3: So I made a commit to remove the two dashes. After doing that, the alignment appears as this now
<img width="1129" alt="Screenshot 2024-12-06 at 12 48 06 PM" src="https://github.com/user-attachments/assets/0aa5931a-3b38-41f7-a6f5-31995a8f9858">

maybe we should add the dashes back

Author: Added them back in, should look good now.

> +
> [source,terminal]
> ----
> oc get managedclusters example-sno -ojsonpath='{.metadata.labels}' | jq
> ----
> 
> View the following example of applied labels:
> 
> .Example applied labels
> [source,json]
> ----
> {
>   "common": "true",
>   "group-du": "",
>   ...
> }
> ----
> 
> * You can verify that your additional annotations are applied by running the following command:
> 
> +
> [source,terminal]
> ----
> oc get bmh example-sno.example.redhat.com -n example-sno -ojsonpath='{.metadata.annotations}' | jq
> ----
> 
> View the following example of applied annotations:
> 
> .Example applied annotation
> [source,json]
> ----
> {
>   "myNodeAnnotation": "success",
>   ...
> }
> ----

Reviewer 3: ```suggestion
+
--
[source,terminal]
----
oc get managedclusters example-sno -ojsonpath='{.metadata.labels}' | jq
----

View the following example of applied labels:

.Example applied labels
[source,json]
----
{
  "common": "true",
  "group-du": "",
  ...
}
----
--

* You can verify that your additional annotations are applied by running the following command:

+
--
[source,terminal]
----
oc get bmh example-sno.example.redhat.com -n example-sno -ojsonpath='{.metadata.annotations}' | jq
----

View the following example of applied annotations:

.Example applied annotation
[source,json]
----
{
  "myNodeAnnotation": "success",
  ...
}
----
--
```

  namespace: "example-sno"
spec:
  [...]
  clusterName: "example-sno"
  extraAnnotations: <1>
    ClusterDeployment:
      myClusterAnnotation: success
  extraLabels: <1>
    ManagedCluster:
      common: "true"
      group-du: ""
  nodes:
    - hostName: "example-sno.example.redhat.com"
      role: "master"
      extraAnnotations: <2>
        BareMetalHost:
          myNodeAnnotation: success
      extraLabels: <2>
        BareMetalHost:
          "testExtraLabel": "success"
----
<1> This field supports cluster-level annotations and labels that the {sco} applies to the the `ManagedCluster` and `ClusterDeployment` manifests.
<2> This field supports node-level annotations and labels that the {sco} applies to the `BareMetalHost` manifest.

* You can verify that your additional labels are applied by running the following command:

+
[source,terminal]
----
oc get managedclusters example-sno -ojsonpath='{.metadata.labels}' | jq
----

View the following example of applied labels:

.Example applied labels
[source,json]
----
{
  "common": "true",
  "group-du": "",
  ...
}
----

* You can verify that your additional annotations are applied by running the following command:

+

> --
> [source,terminal]
> ----
> oc get bmh example-sno.example.redhat.com -n example-sno -ojsonpath='{.metadata.annotations}' | jq
> ----
> 
> View the following example of applied annotations:
> 
> .Example applied annotation
> [source,json]
> ----
> {
>   "myNodeAnnotation": "success",
>   ...
> }
> ----
> --

Reviewer 3: ```suggestion
[source,terminal]
----
oc get bmh example-sno.example.redhat.com -n example-sno -ojsonpath='{.metadata.annotations}' | jq
----

View the following example of applied annotations:

.Example applied annotation
[source,json]
----
{
  "myNodeAnnotation": "success",
  ...
}
----
```

  namespace: example-namespace
----

If the {sco} deletes the resources, `KlusterletAddonConfig` and `ManagedCluster` custom resources are the first to be deleted, while the `AgentClusterInstall` and `ClusterDeployment` custom resources are the last.

[#extra-annotations-labels]
== Configuration of additional annotations and labels

You can configure additional annotations and labels to both cluster-level and node-level installation manifests by using the `extraAnnotations` and `extraLabels` fields in the `ClusterInstance` API.
The {sco} applies your additional annotations and labels to the manifests that you specify in the `ClusterInstance` resource.

When creating your additional annotations and labels, you must specify a manifest type to allow the {sco} to apply them to all the matching manifests.
However, the annotations and labels are arbitrary and you can set any key and value pairs that are meaningful to your applications.

*Note:* The additional annotations and labels are only applied to the resources that were rendered through the referenced templates.

> View the following example application of extraAnnotations and extraLabels:

Reviewer 3: @amolnar-rh will you monospace the parameters here?
